### PR TITLE
feat: add Traversable type class #1582

### PR DESCRIPTION
**Pull request by stephentetley**
This PR adds a Traversable type class an instances for List, Map, Option, Nel, Chain and RedBlackTree. It also adds a "shape agnostic" Eq instance to RedBlackTree - this instance makes tests easier to write.

The Traversable class has just two methods `sequence` and `traverse`. Haskell's Traversable also has `mapM` and `sequenceM` (called `sequence` - our `sequence` is Haskell's `sequenceA`). Haskell's additions give the actions monadic types rather than applicative ones - I think this is largely a legacy decision from when `Applicative` wasn't a superclass of `Monad` so I haven't included them here.

Also added is a Traversable module of 3 functions. `for` is a flipped version of `traverse` - it can be useful if you want a notation like a traditional for loop. The other two functions are `mapAccumLeft` and `mapAccumRight` - these are like a combination of `map` and the respective folds. They've been implemented with local definitions of the State monad (actually just the State applicative as we don't need `flatMap`). The implementations are ContState applicatives rather than simpler State applicatives so they can be used on large structures without blowing the stack (they've been tested on 2 million element lists).

I've put the tests in a single module like SemiGroup does, rather than add them to each data structure's test module.

For `Option`, the already defined `sequence` and `traverse` functions differ from the ones supplied by the Traversable instance (see issue #2747). The existing (module) definitions apply Option as an applicative functor to a list, whereas the correct definitions for a Traversable instance treat `Option` as a _container_ and allow any applicative functor to be applied. As they are different functions (and `Result` and `Validation` also provide equivalent ones), the module definitions seem better to leave as they are for the time being.  

Some of the implementations use a helper function like `consA` in List. These are written in the applicative style but we currently don't have operators defined for `Functor.map` and `Applicative.ap` so they look more unusual than they otherwise would.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Yay! Looking forward to reviewing it :)

**Comment by mlutze** [comment]
This will be a welcome addition to the library!

I have a couple of high-level questions when comparing your implementation to Haskell's:

1. I notice that Haskell's `Traversable` class has a `Foldable` and a `Functor` super class, but this implementation only has a `Functor` super class. Why is that?
2. You've only included a couple of `identity` laws in the class, but the Haskell documentation includes [a few more](https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-Traversable.html#laws). Is there a reason to leave them out? (We're aware of some [issues with type generalization in laws](https://github.com/flix/flix/issues/2099))

**Comment by stephentetley** [comment]
I'd overlooked that `Foldable` was a superclass of `Traversable` in Haskell (and PureScript). `Traversable` in Haskell needs `Foldable` so it can provide `foldMapDefault` (which I don't think we should follow, at least at the moment), but it also indicates the relationship that all `Traversables` are `Foldable` which is valuable - I will add it as a superclass to the PR. 

This Reddit thread was very pro the position to keep `Foldable` a superclass of `Traversable`:

https://www.reddit.com/r/haskell/comments/7dmjh8/why_does_traversable_need_foldable/

Laws - the two identity laws I've implemented I cribbed from elsewhere (my first time trying to code laws in Flix). I can try doing the other laws but it may take some time.

**Comment by magnus-madsen** [comment]
Reddit thread is interesting. Thanks for the link!

**Comment by stephentetley** [comment]
I've added `Foldable` superclass. This needed `Foldable` instances for `Map` and `RedBlack` tree but they already had `foldLeft` / `foldRight` functions so the instances are simple.

**Comment by stephentetley** [comment]
I'm struggling trying to write the other laws.

This description has helped a lot for naturality:

https://stackoverflow.com/questions/59020198/what-does-the-naturality-law-for-traversables-mean

However I don't know how to represent the different instantiations of the natural transformation, I'm stuck at this point - just saying they are the same in a comment:

~~~

    ///
    /// sequence naturality.
    ///
    /// t1 and t2 are different instantiations of the same natural transformation
    ///
    // law naturality: forall(t: t[a], t1: m1[a] -> m2[a], t2: m1[t[a]]] -> t[m2[a]]) with Eq[m2[t[a]]], Applicative[m1], Applicative[m2] . t2(Traversable.sequence(t)) == Traversable.sequence(Functor.map(t1, t))


    ///
    /// traverse naturality.
    ///
    /// t1 and t2 are different instantiations of the same natural transformation
    ///
    law naturality: forall(t: t[a], f: a -> m1[a], t1: m1[a] -> m2[a], t2: m1[t[a]] -> m2[t[a]]) with Eq[m2[t[a]]], Applicative[m1], Applicative[m2] . t2(Traversable.traverse(f, t)) == Traversable.traverse(x -> t1(f(x)), t)

~~~

**Comment by stephentetley** [comment]
I'm stuck with the laws - shall I make the class "lawless" provisionally so the PR can go through?

Also I noticed that PureScript's `Traversable` module also defines `scanLeft` and `scanRight` along with `mapAccumLeft` and `mapAccumRight`, however the scans are slightly different to what we have for the List module - they do not include the initial "state" value - so I don't think we should add them. With just `Traversable` we cannot code versions equivalent to the List module as we do not have "cons". 

https://pursuit.purescript.org/packages/purescript-foldable-traversable/5.0.1/docs/Data.Traversable#t:Traversable

**Comment by magnus-madsen** [comment]
> I'm stuck with the laws - shall I make the class "lawless" provisionally so the PR can go through?

Yes, no worries. We can always add those later. Same for many other classes. Incrementalism 👍 

I have just been too busy to review the PR. Give a me a few days or more :)

File: main/src/library/Map.flix [hunk 1]
```diff
@@ -51,6 +51,22 @@ instance Order[Map[k, v]] with Order[k], Order[v] {
...
+instance Foldable[Map[k]] {
+    pub def foldLeft(f: (b, v) -> b & ef, s: b, m: Map[k, v]): b & ef = Map.foldLeft((acc, v) -> f(acc, v), s, m)
+    pub def foldRight(f: (v, b) -> b & ef, s: b, m: Map[k, v]): b & ef = Map.foldRight((v, acc) -> f(v, acc), s, m)
+}
+
+instance Traversable[Map[k]] {
+    pub def traverse(f: a -> m[b] & ef, t: Map[k, a]): m[Map[k, b]] & ef with Applicative[m] = Map.traverse(f, t)
+    pub override def sequence(t: Map[k, m[a]]): m[Map[k, a]] with Applicative[m] = Map.sequence(t)
+}
+
```
Comment (magnus-madsen) [inline main/src/library/Map.flix:69]: All of the above instances allow me to map/fold/traverse on the value part of a map, right? Keeping the keys the same?

File: main/src/library/Traversable.flix [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        Traversable.traverse(f, t)
+
+    ///
+    /// Returns the result of applying `f` to the traversable structure `t` and the initial state `acc`.
+    /// The result is a pair of the final state and the updated copy of the structure.
+    ///
+    /// `mapAccumLeft` is essentially the combination of `map` and `foldLeft` - like map it returns an updated copy
+    /// of the intial structure, like `foldLeft` it passes an updating accumulator through each step of the traversal.
+    ///
+    pub def mapAccumLeft(f: acc -> a -> (acc, b) & ef, start: acc, t: t[a]): (acc, t[b]) & ef with Traversable[t] =
```
Comment (magnus-madsen) [inline main/src/library/Traversable.flix:65]: Usually we would write the type of `f` as `(acc, a) -> (acc, b)` even though its actually type is the one you have written. Flix is a bit weird in this way. I think the original intention was to "help" the programmer by making it clear that `f` expects two argument (yet still allow it to be used curried).

File: main/src/library/Traversable.flix [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    ///
+    pub def for(t: t[a], f: a -> m[b] & ef): m[t[b]] & ef with Applicative[m], Traversable[t] =
+        Traversable.traverse(f, t)
+
+    ///
+    /// Returns the result of applying `f` to the traversable structure `t` and the initial state `acc`.
+    /// The result is a pair of the final state and the updated copy of the structure.
+    ///
+    /// `mapAccumLeft` is essentially the combination of `map` and `foldLeft` - like map it returns an updated copy
+    /// of the intial structure, like `foldLeft` it passes an updating accumulator through each step of the traversal.
```
Comment (magnus-madsen) [inline main/src/library/Traversable.flix:63]: Thanks for the explanation!

File: main/src/library/Traversable.flix [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    /// ContStateLeft the type of the State monad implemented in CPS. In this case we just need
+    /// `Applicative.ap` so no Monad instance is defined.
+    ///
+    /// A simple implementation of the State monad bursts the stack when `mapAccumLeft` is run on large
+    /// collections, so use ContState the CPS-State monad instead.
+    ///
+    /// We use an Impure continuation so we can internally cast the `ef` of `mapAccumLeft` and not lose
+    /// impurity information.
+    ///
+    opaque type ContStateLeft[ka, s, a] = (a -> s -> ka & Impure) -> s -> ka & Impure
```
Comment (magnus-madsen) [inline main/src/library/Traversable.flix:80]: What's the reason that the function must be `& Impure`. 

Usually functions are either pure or effect polymorphic. I suspect that ContStateLeft should have an extra `ef` parameter. Would that work?

**Review by magnus-madsen — COMMENTED** [review]
Very good.

My only concern is the use of impure functions. I added some notes.

File: main/src/library/Map.flix (hunk 1 above)
Comment (stephentetley) [reply main/src/library/Map.flix:69]: Yes - keys are kept the same (altering could change the "shape" of the map)

File: main/src/library/Traversable.flix (hunk 2 above)
Comment (stephentetley) [reply main/src/library/Traversable.flix:65]: Thanks - I'll change them.

File: main/src/library/Traversable.flix (hunk 4 above)
Comment (magnus-madsen) [reply main/src/library/Traversable.flix:80]: I think you should be able to do that now, but you have to annotate the type parameters, e.g. `enum Foo[a: Type, ef: Bool]`.

**Comment by stephentetley** [comment]
I'm still struggling to write a function monad with a polymorphic effect as a parameter rather than use Impure internally.

With a regular (not CPS) Reader monad - the simplest function monad - I can write the Functor-Applicative-Monad methods as functions with the correct signatures but I can't use them from the instances.

The code below compiles but if I uncomment one or more of the instances it fails type checking.

~~~


namespace Monad {

    pub opaque type ReaderEff[ef: Bool, i: Type, a: Type] = i -> a & ef


/*
    instance Functor[ReaderEff[ef, i]] { 
        pub def map(f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = 
            Monad/ReaderEff.map(f, x)
    }

    instance Applicative[ReaderEff[i]] { 
        pub def point(x: a): ReaderEff[i, a] = 
            Monad/ReaderEff.point(x)

        pub def ap(f: ReaderEff[i, a -> b & ef], x: ReaderEff[i, a]): ReaderEff[i, b] & ef = 
            Monad/ReaderEff.ap(f, x)
    }

    instance Monad[ReaderEff[i]] { 
        pub def flatMap(k: a -> ReaderEff[i, b] & ef, x: ReaderEff[i, a]): ReaderEff[i, b] & ef = 
            Monad/ReaderEff.flatMap(k, x)
    }

*/

}



namespace Monad/ReaderEff {

    use Monad.ReaderEff;
    use Monad.ReaderEff.{ReaderEff};


    pub def map(f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = 
        let ReaderEff(f1) = x;
        ReaderEff(r -> {
            let a = f1(r); 
            f(a)
        }) as & ef



    pub def point(x: a): ReaderEff[ef, i, a] = ReaderEff(_ -> x as & ef)

    pub def ap(f: ReaderEff[ef, i, a -> b & ef], x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = 
        let ReaderEff(f1) = f;
        let ReaderEff(x1) = x;
        ReaderEff(r -> {
            let f = f1(r);
            let a = x1(r);
            f(a)
        }) as & ef 
    

    pub def flatMap(k: a -> ReaderEff[ef, i, b] & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = 
        let ReaderEff(x1) = x;
        ReaderEff(r -> {
            let a = x1(r);
            let ReaderEff(f) = k(a);
            f(r) 
        }) as & ef


    pub def ask(): ReaderEff[ef, i, i] = 
        ReaderEff(r -> r as & ef)

    pub def runReaderEff(x: ReaderEff[ef, i, a], r: i): a & ef = 
        let ReaderEff(x1) = x;
        x1(r)

}

~~~

Failure from uncommenting just the Functor instance: 

~~~

-- Instance Error -------------------------------------------------- ..\flix-monadlib\src\Doodle\ReaderEff.flix


Mismatched signature 'map' required by class 'Functor'.

9 |         pub def map(f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = 
                    ^^^
                    mismatched signature.

Expected scheme: ∀(ef, i, a, ef, b). a -> b & ef -> Monad.ReaderEff[ef, i, a] -> Monad.ReaderEff[ef, i, b] & ef with Functor[Monad.ReaderEff[ef, i]]
Actual scheme:   ∀(a, b, ef, i). a -> b & ef -> Monad.ReaderEff[ef, i, a] -> Monad.ReaderEff[ef, i, b] & ef with Functor[Monad.ReaderEff[ef, i]]
Tip: Modify the definition to match the signature.

Compilation failed with 1 error(s).

Process finished with exit code 1

~~~

File: main/src/library/Traversable.flix (hunk 4 above)
Comment (stephentetley) [reply main/src/library/Traversable.flix:80]: I'm still struggling to write monads with an `ef` parameter - see long note below.

**Comment by magnus-madsen** [comment]
Hmm. Looks like a bug in Flix.

```
Mismatched signature 'map' required by class 'Functor'.

9 |         pub def map(f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = 
                    ^^^
                    mismatched signature.

Expected scheme: ∀(ef, i, a, ef, b). a -> b & ef -> Monad.ReaderEff[ef, i, a] -> Monad.ReaderEff[ef, i, b] & ef with Functor[Monad.ReaderEff[ef, i]]
Actual scheme:   ∀(a, b, ef, i). a -> b & ef -> Monad.ReaderEff[ef, i, a] -> Monad.ReaderEff[ef, i, b] & ef with Functor[Monad.ReaderEff[ef, i]]
```

I note that the expected scheme has two type variables called `ef` (or maybe its our broken pretty printing).

What happens if you try:

```
pub def map[a: Type, b: Type, i: Type, ef: Bool](f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef
```

**Comment by magnus-madsen** [comment]
Minimal broken program:

```
namespace Monad {

    pub opaque type ReaderEff[ef: Bool, i: Type, a: Type] = i -> a & ef

    instance Functor[ReaderEff[ef, i]] { 
        pub def map[a: Type, b: Type, i: Type, ef: Bool](f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = ??? as & ef
    }
}
```

@mlutze Maybe you have some insight on this?

**Comment by stephentetley** [comment]
> What happens if you try:

~~~

pub def map[a: Type, b: Type, i: Type, ef: Bool](f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef

~~~

Output:

~~~

-- Instance Error -------------------------------------------------- ..\flix-monadlib\src\Doodle\ReaderEff.flix


Mismatched signature 'map' required by class 'Functor'.

8 |         pub def map[a: Type, b: Type, i: Type, ef: Bool](f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = 
                    ^^^
                    mismatched signature.

Expected scheme: ∀(ef, i, a, ef, b). a -> b & ef -> Monad.ReaderEff[ef, i, a] -> Monad.ReaderEff[ef, i, b] & ef with Functor[Monad.ReaderEff[ef, i]]
Actual scheme:   ∀(a, b, i, ef, ef, i). a -> b & ef -> Monad.ReaderEff[ef, i, a] -> Monad.ReaderEff[ef, i, b] & ef with Functor[Monad.ReaderEff[ef, i]]
Tip: Modify the definition to match the signature.

Compilation failed with 1 error(s).

Process finished with exit code 1

~~~

**Comment by magnus-madsen** [comment]
OK, I suggest we merge this with the impure casts and then I will open a ticket for the issue. Sounds good?

**Comment by stephentetley** [comment]
> OK, I suggest we merge this with the impure casts and then I will open a ticket for the issue. Sounds good?

Yes - sounds good to me, thanks Magnus.

File: main/src/library/Traversable.flix (hunk 2 above)
Comment (stephentetley) [reply main/src/library/Traversable.flix]: Changed

**Comment by magnus-madsen** [comment]
Thanks Stephen 🎉

**Comment by mlutze** [comment]
> Minimal broken program:
> 
> ```
> pub enum ReaderEff[ef: Bool, i: Type, a: Type] {
>     case ReaderEff(i -> a & ef)
> }
> 
> instance Functor[ReaderEff[ef, i]] { 
>     pub def map[a: Type, b: Type, i: Type, ef: Bool](f: a -> b & ef, x: ReaderEff[ef, i, a]): ReaderEff[ef, i, b] & ef = ??? as & ef
> }
> ```
> 
> @mlutze Maybe you have some insight on this?

I think it might be because the`Functor` class expects the `ef` of the function to be independent from the functor argument `x`, but since `ef` is also in the type of `x`, it fails to unify.

**Pull request merged**